### ISSUE TITLE
Rootful Podman machine cannot bind to a host IP

### ISSUE THREAD

**Issue by shanesmith** [containers/podman#13543]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

A container attempting to bind to a host IP results in an error in a rootful Podman machine.

**Steps to reproduce the issue:**

1. `podman machine init --rootful --now`

2. `sudo ifconfig lo0 alias 192.168.64.1` (creates an alias to the loopback device) 

3. `podman run -it -p 192.168.64.1:8888:8888 alpine`

**Describe the results you received:**

The container does not start and the following error is displayed:

```
Error: error preparing container e9accb477e9dfffd8790b5adec96fb031e7b82beeb60403ba647a2d1f75403be for attach: cannot listen on the TCP port: listen tcp4 192.168.64.1:8888: bind: cannot assign requested address
```

**Describe the results you expected:**

A running container.

**Additional information you deem important (e.g. issue happens only occasionally):**

This happens on a fresh machine, so restarting/recreating doesn't solve it.

Tried with a variety of IP addresses without success.

The following two scenarios do work fine, however:

- `podman run -it -p 8888:8888 alpine` (no host IP seems ok)

- `podman run -it -p 127.0.0.1:8888:8888 alpine` (binding to 127.0.0.1 seems ok)

- `podman -c podman-machine-default run -it -p 192.168.64.1:8888:8888 alpine` (rootless Podman machine seems ok)

**Output of `podman version`:**

```
Client:       Podman Engine
Version:      4.0.2
API Version:  4.0.2
Go Version:   go1.17.8

Built:      Wed Mar  2 09:04:36 2022
OS/Arch:    darwin/arm64

Server:       Podman Engine
Version:      4.0.2
API Version:  4.0.2
Go Version:   go1.16.14

Built:      Thu Mar  3 09:58:50 2022
OS/Arch:    linux/arm64
```

**Output of `podman info --debug`:**

```
host:
  arch: arm64
  buildahVersion: 1.24.1
  cgroupControllers:
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.0-2.fc35.aarch64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.0, commit: '
  cpus: 1
  distribution:
    distribution: fedora
    variant: coreos
    version: "35"
  eventLogger: journald
  hostname: localhost.localdomain
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
    uidmap:
    - container_id: 0
      host_id: 501
      size: 1
    - container_id: 1
      host_id: 100000
      size: 1000000
  kernel: 5.15.18-200.fc35.aarch64
  linkmode: dynamic
  logDriver: journald
  memFree: 1678942208
  memTotal: 2048237568
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.4.2-1.fc35.aarch64
    path: /usr/bin/crun
    version: |-
      crun version 1.4.2
      commit: f6fbc8f840df1a414f31a60953ae514fa497c748
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/501/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.12-2.fc35.aarch64
    version: |-
      slirp4netns version 1.1.12
      commit: 7a104a101aa3278a2152351a082a6df71f57c9a3
      libslirp: 4.6.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.3
  swapFree: 0
  swapTotal: 0
  uptime: 4m 26.02s
plugins:
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /var/home/core/.config/containers/storage.conf
  containerStore:
    number: 8
    paused: 0
    running: 0
    stopped: 8
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/home/core/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 2
  runRoot: /run/user/501/containers
  volumePath: /var/home/core/.local/share/containers/storage/volumes
version:
  APIVersion: 4.0.2
  Built: 1646319530
  BuiltTime: Thu Mar  3 09:58:50 2022
  GitCommit: ""
  GoVersion: go1.16.14
  OsArch: linux/arm64
  Version: 4.0.2
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
$ brew info podman
podman: stable 4.0.2 (bottled), HEAD
Tool for managing OCI containers and pods
https://podman.io/
/opt/homebrew/Cellar/podman/4.0.2 (172 files, 47.9MB) *
  Poured from bottle on 2022-03-17 at 10:50:58
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/podman.rb
License: Apache-2.0
==> Dependencies
Build: go ✔, go-md2man ✔
Required: qemu ✔
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

✅ Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

MacBook Pro M1

**Comment by mheon** [comment]
Any chance you can provide a `podman inspect` of a container which shows this symptom? Want to make sure we're just binding to that one IP.

**Comment by shanesmith** [comment]
@mheon The container never does start due to the error. I've updated the OP to mention this now.

**Comment by Luap99** [comment]
I think I know the problem. We try to bind the ports as root in libpod but I don't call the magic machine port conversion function there.

**Comment by Luap99** [comment]
Should be fixed by https://github.com/containers/podman/pull/13544

**Comment by Luap99** [comment]
For this patch you only have to change the podman binary in the VM.
Since the VM is fedora35 you can copy the binary from the CI build https://cirrus-ci.com/task/4863906593112064, just click on binary under the Artifacts section. You need the "podman" binary there.
In your VM activate rpm-ostree usroverlay with `podman machine ssh sudo rpm-ostree usroverlay`, then you should be able to copy the new binary to `/usr/bin/podman` in the VM. This will make the machine use the new binary. Note that this change is only active until the next reboot of the VM.

**Comment by Luap99** [comment]
Ah yeah, we do have arm systems in CI unfortunately. You would need to compile from source then.

**Comment by shanesmith** [comment]
Verified fixed with a build from source, thanks!

**Comment by shanesmith** [comment]
Would I be right in assuming this fix will make it into v4.0.3? =)

**Comment by mheon** [comment]
Assuming it backports cleanly, yes
